File: src/dbus.rs
use std::collections::HashMap;
use std::convert::TryInto;

use hex::FromHex;
use zbus::export::zvariant;
use zbus::{dbus_interface, fdo, Connection, ObjectServer};

use crate::{storage::DocumentId, Domain};

const SERVER_NAME: &str = "com.github.pvdrz.domain";
const SERVER_PATH: &str = "/com/github/pvdrz/domain";

pub(crate) fn serve(domain: Domain) -> fdo::Result<()> {
    let connection = Connection::new_session()?;

    fdo::DBusProxy::new(&connection)?
        .request_name(SERVER_NAME, fdo::RequestNameFlags::ReplaceExisting.into())?;

    let mut object_server = ObjectServer::new(&connection);
    object_server.at(&SERVER_PATH.try_into().unwrap(), domain)?;

    println!("looping");
    loop {
        object_server.try_handle_next()?;
    }
}

#[dbus_interface(name = "org.gnome.Shell.SearchProvider2")]
impl Domain {
    fn get_initial_result_set(&self, terms: Vec<&str>) -> Vec<String> {
        let query = terms.join(" ");
        self.search(&dbg!(query))
            .into_iter()
            .map(|DocumentId(bytes)| hex::encode(bytes))
            .collect()
    }

    fn get_subsearch_result_set(
        &self,
        _previous_results: Vec<&str>,
        terms: Vec<&str>,
    ) -> Vec<String> {
        self.get_initial_result_set(terms)
    }

    fn get_result_metas(
        &self,
        str_ids: Vec<String>,
    ) -> Vec<HashMap<&'static str, zvariant::Value>> {
        let mut metas = Vec::with_capacity(str_ids.len());

        for str_id in str_ids {
            let id = match <[u8; std::mem::size_of::<u64>()]>::from_hex(&str_id) {
                Ok(bytes) => DocumentId(bytes),
                Err(_) => panic!(),
            };

            let doc = match self.get(id) {
                Ok(Some(doc)) => doc,
                Ok(None) | Err(_) => panic!(),
            };

            let meta = {
                let mut meta = HashMap::with_capacity(3);
                meta.insert("id", str_id.into());
                meta.insert("name", doc.title.into());
                meta.insert("description", doc.authors.join(", ").into());
                meta
            };

            metas.push(meta);
        }

        metas
    }

    fn activate_result(&self, str_id: &str, _terms: Vec<&str>, _timestamp: u32) {
        let id = DocumentId(<[u8; std::mem::size_of::<u64>()]>::from_hex(str_id).unwrap());

        let doc = self.get(id).unwrap().unwrap();

        let path = self
            .config
            .path
            .join(hex::encode(doc.hash))
            .with_extension(doc.extension);

        open::that(dbg!(path)).unwrap();
    }

    fn launch_search(&self, _terms: Vec<&str>, _timestamp: u32) {}
}

File: cmd/serve/dbus.go
package main

import (
	"encoding/hex"
	"errors"
	"path"
	"strings"

	"github.com/godbus/dbus/v5"
	"github.com/pvdrz/domain/lib/doc"
	"github.com/skratchdot/open-golang/open"
)

const serverName = "com.github.pvdrz.domain"
const serverPath = "/com/github/pvdrz/domain"

func ServeDbus(domain Domain) error {
	conn, err := dbus.ConnectSessionBus()
	if err != nil {
		return err
	}
	defer conn.Close()

	conn.Export(&domain, serverPath, "org.gnome.Shell.SearchProvider2")

	reply, err := conn.RequestName(serverName, dbus.NameFlagDoNotQueue)
	if err != nil {
		return err
	}

	if reply != dbus.RequestNameReplyPrimaryOwner {
		return errors.New("name already taken")
	}

	select {}
}

func (domain *Domain) GetInitialResultSet(terms []string) ([]string, *dbus.Error) {
	query := strings.Join(terms, " ")
	ids := domain.search(query)

	results := make([]string, len(ids))
	for i, id := range ids {
		results[i] = id.ToString()
	}

	return results, nil
}

func (domain *Domain) GetSubsearchResultSet(_ []string, terms []string) ([]string, *dbus.Error) {
	return domain.GetInitialResultSet(terms)
}

func (domain *Domain) GetResultMetas(strIDs []string) ([]map[string]dbus.Variant, *dbus.Error) {
	metas := make([]map[string]dbus.Variant, len(strIDs))

	for i, strID := range strIDs {
		id, err := doc.DocIDFromString(strID)
		if err != nil {
			return metas, dbus.MakeFailedError(err)
		}

		doc, err := domain.get(id)
		if err != nil {
			return metas, dbus.MakeFailedError(err)
		}

		meta := make(map[string]dbus.Variant, 3)
		meta["id"] = dbus.MakeVariant(strID)
		meta["name"] = dbus.MakeVariant(doc.Title)
		meta["description"] = dbus.MakeVariant(strings.Join(doc.Authors, ", "))

		metas[i] = meta
	}

	return metas, nil
}

func (domain *Domain) ActivateResult(strID string, _ []string, _ uint32) *dbus.Error {
	id, err := doc.DocIDFromString(strID)
	if err != nil {
		return dbus.MakeFailedError(err)
	}

	doc, err := domain.get(id)
	if err != nil {
		return dbus.MakeFailedError(err)
	}

	filename := hex.EncodeToString(doc.Hash[:]) + "." + doc.Extension
	path := path.Join(domain.config.Path, filename)
	err = open.Start(path)
	if err != nil {
		return dbus.MakeFailedError(err)
	}

	return nil
}

func (domain *Domain) LaunchSearch(_ []string, _ uint32) *dbus.Error {
	return nil
}
